### ISSUE TITLE
Is it possible to open and ignore (or warn) rather than fail if features have None geometry?

### ISSUE THREAD

**Issue by wheeled** [Toblerity/Fiona#789]
## Expected behavior and actual behavior.

I was unable to open a .GPKG file from abs.gov.au because it contains features that do not have geometry.  The traceback looks like this:
```
Traceback (most recent call last):
  File "/Users/wheeled/PycharmProjects/misc37/language/language.py", line 29, in <module>
    gdf = gpd.read_file(filepath, driver="GPKG")  #layer='census2016_cldg_nsw_ssc_short'
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/geopandas/io/file.py", line 95, in read_file
    columns = list(features.meta["schema"]["properties"]) + ["geometry"]
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/fiona/collection.py", line 228, in meta
    'driver': self.driver, 'schema': self.schema, 'crs': self.crs,
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/fiona/collection.py", line 206, in schema
    self._schema = self.session.get_schema()
  File "fiona/ogrext.pyx", line 661, in fiona.ogrext.Session.get_schema
  File "fiona/_geometry.pyx", line 78, in fiona._geometry.normalize_geometry_type_code
fiona.errors.UnsupportedGeometryTypeError: 3000
```

It took a while to track back to find that the fatal error occurs in collection.py at line 158 where the ogr Session is initialised.  

```
        try:
            if self.mode == 'r':
                self.session = Session()
                print(self.meta).  # line added to debug
                self.session.start(self, **kwargs)
            elif self.mode in ('a', 'w'):
                self.session = WritingSession()
                self.session.start(self, **kwargs)
        except IOError:
            self.session = None
            raise
```

After initialisation, the 'meta' property produces an exception rather than a dict.  This exception is masked by the try/except block which only looks for an IOError.  It can be exposed by attempting to access the meta property within the try/except block:

```
Traceback (most recent call last):
  File "/Users/wheeled/PycharmProjects/misc37/language/language.py", line 29, in <module>
    gdf = gpd.read_file(filepath, driver="GPKG")  #layer='census2016_cldg_nsw_ssc_short'
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/geopandas/io/file.py", line 77, in read_file
    with reader(path_or_bytes, **kwargs) as features:
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/fiona/env.py", line 397, in wrapper
    return f(*args, **kwargs)
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/fiona/__init__.py", line 253, in open
    layer=layer, enabled_drivers=enabled_drivers, **kwargs)
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/fiona/collection.py", line 159, in __init__
    print(self.meta)
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/fiona/collection.py", line 228, in meta
    'driver': self.driver, 'schema': self.schema, 'crs': self.crs,
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/fiona/collection.py", line 206, in schema
    self._schema = self.session.get_schema()
  File "fiona/ogrext.pyx", line 593, in fiona.ogrext.Session.get_schema
ValueError: Null layer
```

I eventually opened the file using ogr itself and found features without geometry which I suspect are producing the exception.  I could not find a kwarg that would allow fiona.open to ignore invalid geometry.  Is there a way?

## Steps to reproduce the problem.

The file producing the issue is 96 MB so I won't upload it, but it can be downloaded from ABS if desired.  The script to produce the problem is very simple (geopandas uses fiona to open the file):

```
import geopandas as gpd
filepath = r'/Users/wheeled/GEOSPATIAL/Geopackage_2016_CLDG_for_NSW/census2016_cldg_nsw_short.gpkg'
gdf = gpd.read_file(filepath, driver="GPKG")
```

Test file if required can eb downloaded from [https://datapacks.censusdata.abs.gov.au/geopackages/](url) with the following drop down selections:
 - New South Wales
 - Cultural and Language Diversity
 - CLD Package G

## Operating system

For example: Mac OS X 10.14.6.

## Fiona and GDAL version and provenance

Libraries all installed using PyCharm 2018.3.6
fiona 1.8.6
GDAL 2.4.0
geopandas 0.5.1

**Comment by sgillies** [comment]
@wheeled I think this has something to do with the way geopandas uses the fiona module instead of being a fiona bug. Fiona can handle null geometries and I successfully dumped all the features of the first layer out to a text file:

```
$ fio --version
fio, version 1.8.6
$ fio cat ~/Downloads/Geopackage_2016_CLDG_for_NSW/census2016_cldg_nsw_short.gpkg --layer 1:census2016_cldg_nsw_ced_short > /tmp/census.jsons
$ grep -c Feature /tmp/census.jsons
49
```

Can you try

```
gdf = gpd.read_file(filepath, driver="GPKG", layer="census2016_cldg_nsw_ced_short")
```

and see if you have better results?

**Comment by wheeled** [comment]
@sgillies I see the same (exception) result when I add the layer= kwarg in geopandas.  I also see the same exception if I use fiona directly to open the gpkg using:
```
import fiona

filepath = r'/Users/wheeled/GEOSPATIAL/Geopackage_2016_CLDG_for_NSW/census2016_cldg_nsw_short.gpkg'

for layername in fiona.listlayers(filepath):
    with fiona.open(filepath, driver="GPKG", layer=layername) as src:
        print(layername, len(src))
```

However, I see the same result as you and the json makes sense when I replicate the CLI invocation you used above.  I need to dig into it some more, but something must be different between fio and fiona?

**Comment by wheeled** [comment]
@sgillies I had a deeper dive today and I believe I know where the issue is.  In `ogrext.pyx` at the end of the `Session.get_schema` method, there is a call to `normalize_geometry_type_code`, which is passed an `ogr.GetGeometryType` value (I'm refering to the Python API).  

```
        if not self.collection.ignore_geometry:
            code = normalize_geometry_type_code(
                OGR_FD_GetGeomType(cogr_featuredefn))
            ret["geometry"] = GEOMETRY_TYPES[code]
```

If the geometry is `None`, then `normalize_geometry_type_code` raises the `UnsupportedGeometryTypeError`.  If the `ignore_geometry` kwarg is `True` then the issue doesn't arise.  

The issue arose in geopandas because it uses `Collection.meta` as previously noted, and this calls `Collection.session.get_schema`.  

I believe there needs to be a fix in either `Session.get_schema` or `normalize_geometry_type_code` to handle the case when the geometry is `None`.  Note that you can't see the problem unless you are using the `schema` property (either directly or via `meta`).  This test script will illustrate:

```
import fiona

for layername in ['census2016_cldg_nsw_ced_short']:
    with fiona.open(
            filepath,
            driver="GPKG",
            layer=layername,
            ignore_geometry=False,
    ) as src:
        print(layername, len(src))
        print(src.schema)
```
which produces:
```
census2016_cldg_nsw_ced_short 49
Traceback (most recent call last):
  File "/Users/wheeled/PycharmProjects/misc37/language/language.py", line 24, in <module>
    print(src.schema)
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/fiona/collection.py", line 205, in schema
    self._schema = self.session.get_schema()
  File "fiona/ogrext.pyx", line 661, in fiona.ogrext.Session.get_schema
  File "fiona/_geometry.pyx", line 78, in fiona._geometry.normalize_geometry_type_code
fiona.errors.UnsupportedGeometryTypeError: 3000
```

Would you agree this is a Fiona bug?  I'm not clear on why the exception code is 3000.

**Comment by sgillies** [comment]
@wheeled thanks for looking deeper into this! I think it's the same bug that was fixed only for the master branch (to be version 1.9) in https://github.com/Toblerity/Fiona/pull/724. I'm going to schedule a 1.8.7 release for tomorrow which should get you unstuck.

**Comment by wheeled** [comment]
@sgillies Thanks - I will update to 1.8.7 and let you know.  Looks like the same issue, and would explain why the exception code is 3000.  Rookie error on my part - I was reading the master version of _geometry.pyx not the 1.8.6 version I had.

**Comment by wheeled** [comment]
Hi @sgillies 
After upgrade to 1.8.7 I am now seeing the following error:
`fiona._err.CPLE_AppDefinedError: unable to open database file: this file is a WAL-enabled database. It cannot be opened because it is presumably read-only or in a read-only directory.
`
Is that likely to be a result of one of the other changes in the upgrade?

**Comment by sgillies** [comment]
@wheeled yes, looks like something has changed between GDAL 2.3.3 and 2.4.2 with respect to WAL mode. We'll need to figure that out because we can't stay at GDAL 2.3.3.

Meanwhile, the fix I made for 1.8.7 has uncovered another bug, so I'm working on 1.8.8.

**Comment by sgillies** [comment]
1.8.8 is on PyPI now.

I can't reproduce the WAL-enabled error using the GPKG file in my Downloads directory. There may be a more subtle issue with permissions on your computer? https://trac.osgeo.org/gdal/ticket/6776 seems to be where that error message was introduced to GDAL.

**Comment by wheeled** [comment]
@sgillies I am not seeing the WAL issue today.  The only thing I am aware of that changed is that Apple did a minor update and my Mac rebooted.  Perhaps the environment was not properly initialised after I upgraded yesterday.

I'm on MacOS 10.14.6 (18G95) fyi.

**HOWEVER** - the geometry type error has reappeared.  The features with the `None` geometry now indicate a `2**31` code:
```
Traceback (most recent call last):
  File "/Users/wheeled/PycharmProjects/misc37/language/language.py", line 52, in <module>
    print(features.schema)
  File "/Users/wheeled/Library/Python/3.6/lib/python/site-packages/fiona/collection.py", line 205, in schema
    self._schema = self.session.get_schema()
  File "fiona/ogrext.pyx", line 661, in fiona.ogrext.Session.get_schema
  File "fiona/_geometry.pyx", line 78, in fiona._geometry.normalize_geometry_type_code
fiona.errors.UnsupportedGeometryTypeError: 2147483648
```
Bit confused here, because I am pretty sure I looked for that in 1.8.7 yesterday.

**Comment by wheeled** [comment]
@ sgillies I just figured out that you must have spotted that and fixed it in 1.8.8, but I haven't upgraded to 1.8.8 yet because I saw another bug report.
